### PR TITLE
added subscription_error_count reset / AMQP IOTHUB_CLIENT_CONNECTION_QUOTA_EXCEEDED

### PR DESCRIPTION
**Pull request by ericwol-msft**
# Checklist
- [x] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-c/blob/main/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [ ] This pull-request is submitted against the `main` branch. 
  - [ ] I have merged the latest `main` branch prior to submission and re-merged as needed after I took any feedback.
  - [ ] I have squashed my changes into one with a clear description of the change.

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

**Comment by ericwol-msft** [comment]
#2249

File: iothub_client/src/iothubtransport_amqp_twin_messenger.c [hunk 1]
```diff
@@ -2005,6 +2005,7 @@ int twin_messenger_stop(TWIN_MESSENGER_HANDLE twin_msgr_handle)
         }
         else
         {
+            twin_msgr->subscription_error_count = 0;
```
Comment (danewalton) [inline iothub_client/src/iothubtransport_amqp_twin_messenger.c:2008]: interesting this wasn't here before. Why is it only now that it's needed? Seems more universally true.

**Review by danewalton — APPROVED** [review]

**Comment by danewalton** [comment]
I'm guessing tests for this would be difficult....

File: iothub_client/src/iothubtransport_amqp_telemetry_messenger.c [hunk 2]
```diff
@@ -981,7 +980,49 @@ static void internal_on_event_send_complete_callback(void* context, MESSAGE_SEND
...
+                                for (uint32_t t = 0; !isResourceLimitExceeded && ret == 0 && t < item_properties_count; t++)
+                                {
+                                    AMQP_VALUE item_property = amqpvalue_get_list_item(item_properties, t);
+                                    if (item_property != NULL)
+                                    {
+                                        const char* symbol_value;
+                                        int ret_sym = amqpvalue_get_symbol(item_property, &symbol_value);
+                                        if (ret_sym == 0)
+                                        {
+                                            if (strcmp(symbol_value, "amqp:resource-limit-exceeded") == 0)
```
Comment (ewertons) [inline iothub_client/src/iothubtransport_amqp_telemetry_messenger.c:1005]: Could you define this as macro on the top and reference it here?
Sugg: 
```c
#define AMQP_ERROR_LIMIT_EXCEEDED "amqp:resource-limit-exceeded"
```

Comment (ewertons) [inline iothub_client/src/iothubtransport_amqp_telemetry_messenger.c:1005]: Could you use `strncmp` instead, as it is safer?

File: iothub_client/src/iothubtransport_amqp_twin_messenger.c (hunk 1 above)
Comment (ewertons) [reply iothub_client/src/iothubtransport_amqp_twin_messenger.c:2008]: The lack of that line was apparently a bug.

Comment (ewertons) [reply iothub_client/src/iothubtransport_amqp_twin_messenger.c:2008]: There was a github issue complaining about that.

**Review by ewertons — DISMISSED** [review]

File: iothub_client/tests/iothubtr_amqp_tel_msgr_ut/iothubtr_amqp_tel_msgr_ut.c [hunk 3]
```diff
@@ -56,6 +56,10 @@ void real_free(void* ptr)
 #define please_mock_amqpvalue_create_string MOCK_ENABLED
```
[the remark this answers is not in the document]
Comment (CIPop) [reply iothub_client/tests/iothubtr_amqp_tel_msgr_ut/iothubtr_amqp_tel_msgr_ut.c:56]: +1 - this requires Unit Testing to pass through the new `AMQP_ERROR_LIMIT_EXCEEDED` case.

File: iothub_client/src/iothubtransport_amqp_telemetry_messenger.c [hunk 4]
```diff
@@ -981,7 +980,49 @@ static void internal_on_event_send_complete_callback(void* context, MESSAGE_SEND
...
+                                uint32_t item_properties_count = 0;
+                                ret = amqpvalue_get_list_item_count(item_properties, &item_properties_count);
+                                for (uint32_t t = 0; !isResourceLimitExceeded && ret == 0 && t < item_properties_count; t++)
+                                {
+                                    AMQP_VALUE item_property = amqpvalue_get_list_item(item_properties, t);
+                                    if (item_property != NULL)
+                                    {
+                                        const char* symbol_value;
+                                        int ret_sym = amqpvalue_get_symbol(item_property, &symbol_value);
+                                        if (ret_sym == 0)
```
Comment (CIPop) [inline iothub_client/src/iothubtransport_amqp_telemetry_messenger.c:1003]: Any reason `ret` wasn't used here? 
What are the error conditions for `ret_sym` being non-zero? Can it indicate memory or data corruption?

File: iothub_client/src/iothubtransport_amqp_telemetry_messenger.c (hunk 2 above)
Comment (CIPop) [reply iothub_client/src/iothubtransport_amqp_telemetry_messenger.c:1005]: We should check for other conditions as well: https://github.com/Azure/azure-iot-sdk-csharp/blob/48163a8f45d6dba186b5937efcf3eab08a7360cd/iothub/device/devdoc/amqpTransportExceptions.md

**Review by CIPop — COMMENTED** [review]
The code should look for other AMQP error codes.
Missing UT.

File: iothub_client/tests/iothubtr_amqp_tel_msgr_ut/iothubtr_amqp_tel_msgr_ut.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply iothub_client/tests/iothubtr_amqp_tel_msgr_ut/iothubtr_amqp_tel_msgr_ut.c:56]: anyone volunteering?

**Pull request merged**